### PR TITLE
 refactor(router): Ensure computed state restoration works for thrown errors

### PR DESCRIPTION
**Pull request by atscott**
When `canceledNavigationResolution='computed'`, the `Router` needs to
handle the cases where errors are thrown. Previously, the logic was not
updated and would simply do a `replaceState` rather than determining
where in the history we should move to restore the page/url from before
the failed navigation.

Note to reviewer: This is marked as a refactor because the feature is not available in the view public API so a line in the CHANGELOG for a bug fix would not make sense.

### CODE REVIEW

File: packages/router/src/router.ts [hunk 1]
```diff
@@ -771,7 +771,8 @@ export class Router {
 
                      filter(t => {
                        if (!t.guardsResult) {
-                         this.cancelNavigationTransitionRestoreHistory(t, '');
+                         this.restoreHistory(t);
+                         this.cancelNavigationTransition(t, '');
```
Comment (AndrewKushnir) [inline packages/router/src/router.ts:775]: nit: it looks like the `cancelNavigationTransition` function is usually called with the preceding `restoreHistory` call (except for 1 case I believe). May be we can move the `restoreHistory` one inside the `cancelNavigationTransition` and have a flag to skip the `restoreHistory` call in that one case? That should help reduce extra code a bit without sacrificing readability/clarity of code.

File: packages/router/src/router.ts [hunk 2]
```diff
@@ -889,7 +891,9 @@ export class Router {
                            // AngularJS sync code which looks for a value here in order to determine
                            // whether or not to handle a given popstate event or to leave it to the
                            // Angular router.
-                           this.cancelNavigationTransitionRestoreHistory(t, cancelationReason);
+                           this.restoreHistory(t);
+                           this.cancelNavigationTransition(
+                               t, `At least one route resolver didn't emit any value.`);
```
Comment (AndrewKushnir) [inline packages/router/src/router.ts:896]: We can probably also move the cancellation message to a const and use it in 2 places (here and in the code above)?

File: packages/router/src/router.ts [hunk 3]
```diff
@@ -908,6 +912,10 @@ export class Router {
                        this.currentNavigation = null;
                      }),
                      catchError((e) => {
+                       // TODO(atscott): The NavigationTransition `t` used here does not accurately
+                       // reflect the current state of the whole transition because some operations
+                       // return a new object rather than modifying the one in the outermost
+                       // `switchMap`.
```
Comment (AndrewKushnir) [inline packages/router/src/router.ts:918]: Quick question: just curious if there is a proposed solution/fix to the problem described in the TODO that we can include here? Or it's unknown atm (needs more investigation)?

File: packages/router/src/router.ts [hunk 4]
```diff
@@ -1472,30 +1482,42 @@ export class Router {
...
+      if (browserUrlUpdateOccurred && targetPagePosition !== 0) {
         this.location.historyGo(targetPagePosition);
+      } else if (
+          this.currentUrlTree === this.currentNavigation?.finalUrl && targetPagePosition === 0) {
+        // We got to the activation stage, but we weren't moving anywhere in history
+        // (skipLocationChange or replaceUrl). We still need to reset the router state back to what
+        // it was when the navigation started.
+        // TODO(atscott): this restore logic is duplicated here and for the `'replace'` resolution
+        // in `resetStateAndUrl`. Investigate moving all of this logic to this function only and
+        // removing `resetStateAndUrl` completely.
```
Comment (AndrewKushnir) [inline packages/router/src/router.ts:1508]: Would it make sense to move it to a helper method for now?

**Review by AndrewKushnir — APPROVED** [review]
LGTM 👍 Just left a few minor comments (nothing critical, just some small proposals).

File: packages/router/src/router.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/router/src/router.ts:775]: 👍

File: packages/router/src/router.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/router/src/router.ts:917]: Ok, thanks for additional info! I was thinking if it makes sense to put some of this info into the comment, so when it's the right time to fix that - there is a context available. Up to you whether it's worth adding that or not :)

Comment (atscott) [reply packages/router/src/router.ts:917]: done

**Comment by atscott** [comment]
[presubmit](https://test.corp.google.com/ui#id=OCL:386505454:BASE:386505486:1627067729052:87180910)

**Pull request closed without merging**